Bad class file magic or version

AbandonedCart opened the pull request:
This resolves a build issue for Android when using this dependency in PokeGOAPI-Java

AbandonedCart committed:
Bad class file magic or version

Paul776 commented:
Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Ljava/util/Base64;

AbandonedCart committed:
Android / Java 7 compliant Base64

Ported from
https://github.com/android/platform_frameworks_base/blob/master/core/jav
a/android/util/Base64.java

mgorski-mg commented on src/main/java/android/util/Base64.java:
```diff
@@ -0,0 +1,741 @@
+ * limitations under the License.
+ */
+
+package android.util;
```
I don't think that it's good idea to use package 'android' in java project. If I'm understand correctly it's Base64 class implementation from android sdk, but please add it to normal package like 'svarzee.gps.gpsoauth.util'
You can also add here implementation from Java8 instead of Android's.

AbandonedCart replied:
It is a 1:1 port of the android framework code, and thus not an extension. It will not override Android platform code, as it is not directly run by an Android package, so it can / should be overridden by the Android local code. This invalidates renaming the package.

Using the implementation from Java 8 requires Java 8, which is the entire reason this change was necessary. Java 7 has a Base64 implementation, but it would not support the requirements for the Java 8 usage in this library, while the Android code is an obvious alternative given the issue originates when using this code with Android.

mgorski-mg replied:
Yup, I know that it's port from Android and it's not the framework dependency. I'm only talking about package naming convention. It's strange when in pure java project we have package 'android'.  And I'm sure that you can rename package name ;)
And obviously I know what the reason is because I have the same problem on Android. I suggested to copy into this library implementation of Base64 from java8. And it will be working too because its only encrypting algorithm. Without any specyfice java8 mechanism.

It's only my recommendation to just copy Java8 code into project and add it to 'svarzee.gps.gpsoauth.util' pacakge. And of course all licence information should be included in Base64 java file too.

AbandonedCart commented:
The naming convention is to retain the original package definition, unless necessary to rename it. It is no less strange to have a package named android that is from Android than it would be to have a package named java when it came from Java.
It sounds like less of a justification and more just a personal opposition to Android. If you have any legitimate issues, I will be more than happy to make adjustments for those. Until then, it is up to the author how this is used, if at all.

mgorski-mg commented:
Yes, that's right that it is up to the author of the library.
For me your solution is good enough. I'm using the same solution so let it to the author.

john086 commented:
Now i have already compatibility error with java 1.7 using instant run

Error:Error converting bytecode to dex:
Cause: Dex cannot parse version 52 byte code.
This is caused by library dependencies that have been compiled using Java 8 or above.
If you are using the 'java' gradle plugin in a library submodule add 
targetCompatibility = '1.7'
sourceCompatibility = '1.7'
to that submodule's build.gradle file.
Error:Execution failed for task ':app:transformClassesWithDexForDebug'.

> com.android.build.api.transform.TransformException: java.lang.RuntimeException: com.android.ide.common.process.ProcessException: java.util.concurrent.ExecutionException: com.android.ide.common.process.ProcessException: Return code 1 for dex process

AbandonedCart commented:
You added a line you didn't need, left out the second commit, and need to set the dependency to your new local build so it isn't ignored by upstream release.

For testing purposes, you can replace `com.github.svarzee:gpsoauth-java:v0.2.0` in the PokeGOAPI-Java build.gradle with `com.github.AbandonedCart:gpsoauth-java:v0.2.2`

john086 commented:
sorry i don't understand, i download local Pokemon PokeGOAPI-Java build it and put it in lib folder of my android project, it's rigth way ?

mgorski-mg commented:
@john086 it can't work with Android right now, it's the reason why we are talking about @LoungeKatt resolution. To make it workable on Android you must build it as java1.7 and replace Base64 implementation because for now this library is using implementation from java 1.8.
To make it working you must download fork from LoungeKatt, compile it to jar. Add this jar to PokeGOAPI-Java/libs, 
replace `compile 'com.github.svarzee:gpsoauth-java:v0.2.0'` from build.gradle  
to `compile fileTree(dir: 'libs', include: ['*.jar'])`

Or just make it simpler as LoungeKatt said.
Just open build.gradle file and chanage `com.github.svarzee:gpsoauth-java:v0.2.0` to `com.github.AbandonedCart:gpsoauth-java:v0.2.2`

By the way, @LoungeKatt are you trying to login after this changes? Because I'm still getting 'BadAuthorization'.

AbandonedCart commented:
They are blocking third-party auth at the moment. It applies to both the automatic and oauth versions, as well as some iPhone official versions.

svarzee commented:
I decided to use a lightweight library instead
http://iharder.sourceforge.net/current/java/base64/

AbandonedCart commented:
Why reference an entire library for a single class? Why not simply pull the class into the local code and spare a huge amount of overhead? While it has a different package, it is the same basic class.
